text/0000-glimmer-components.md

```diff
@@ -0,0 +1,835 @@
+
+```js
+let { firstNode, lastNode } = bounds;
+let currentNode = firstNode;
+
+while (currentNode !== lastNode) {
+  // do something with current node
+  currentNode = currentNode.nextSibling;
+}
+```
+
+This exposes users to low level APIs which may be useful to some, but in general
+are too complicated and nuanced. Instead, we can use new element modifiers which
+allow users to run code declaratively on the element:
+
+```hbs
+<div {{on-insert this.didInsertElement}} ...attributes>
+  First root element
+</div>
+<div {{on-insert this.didInsertElement}}>
+  Second root element
+</div>
+```
+
+This allows users to target which elements they care about directly. It also
+allows for code to be run on elements which are contained in conditionals,
+meaning users do not need to add messy logic to track component state
+themselves:
+
+```hbs
+{{#if shouldRender}}
+  <div
+    {{on-insert this.didInsertElement}}
+    {{on-destroy this.willDestroyElement}}
+  >
+
+  </div>
+{{/if}}
+```
+
+Finally, it means that users can bypass component code altogether by using
+helpers:
+
+```hbs
+<div {{on-insert (setScrollPosition @savedScrollPosition)}}></div>
+```
+
+This RFC proposes adding two element modifiers:
+
+1. `{{on-insert}}`, which runs immediately after the element is inserted.
+   `{{on-insert}}` is guaranteed to run before the enclosing component's
+   `didCreate` and `didUpdate` hooks, and after any child element's
+   `{{on-insert}}` modifiers, preserving the semantics of `didInsertElement`.
+
+   `{{on-insert}}` receives a function with the following signature:
+
+   ```ts
+   type OnInsertHandler = function(element: Element): void;
+   ```
+
+2. `{{on-destroy}}`, which runs immediately before the element is removed.
+   `{{on-destroy}}` is guaranteed to run before the enclosing component's
+   `didUpdate` and `destroy` hooks, and after and child element's
+   `{{on-destroy}}` modifiers, preserving the semantics of `willDestroyElement`.
+
+   `{{on-destroy}}` receives a function with the following signature:
+
+   ```ts
+   type OnDestroyHandler = function(element: Element): void;
+   ```
+
+#### Default Binding of Functions
+
+Currently, the `{{action}}` modifier is the only modifier which ships with
+Ember, and `{{action}}` by default binds functions to the context of the
+_component_. With nested values, this behavior does not particularly make sense:
+
+```hbs
+<div {{on-insert this.animationService.registerElement}}></div>
+```
+
+In this example, `on-insert` would effectively be running
+`this.animationService.registerElement.call(this, element)`, where `this` refers
+to the _component_ and not the _animation service_.
+
+The `bind` RFC, which runs parallel to this one, proposes that a new helper with
+the opposite semantics be introduced. If the `bind` RFC is accepted before this
+RFC is implemented, then the default behavior when using `on-insert` should
+match with the `bind` helper, and _not_ the `action` helper.
```

Comment:
Q: would landing the `bind` helper after the fact, "upgrade" this behavior? Or will that ship have sailed, if so do we want a dependency or is it fine to proceed in any order.

Reply:
There are a few options here:

1. Modifiers could _not_ bind, so `{{on-insert (action this.didInsert)}}` would be required.
2. Modifiers could throw if they detect that they are chaining, so `{{on-insert this.foo.didInsert}}` would throw, but `{{on-insert this.didInsert}}` would not.
3. Modifiers can adopt `{{action}}`s binding mechanics.

Options 1 and 2 allow us to define this behavior later, option 3 locks us in. I think 1 and 2 would both be acceptable to remove the dependency